REGRESSION(276827@main) Retain frame request queue in ImageFrameWorkQueue::start

achristensen07 opened the pull request:
#### 0b216b7fee1af5547c9b478998c8ea3c412e6489
<pre>
REGRESSION(276827@main) Retain frame request queue in ImageFrameWorkQueue::start
<a href="https://bugs.webkit.org/show_bug.cgi?id=271979">https://bugs.webkit.org/show_bug.cgi?id=271979</a>
<a href="https://rdar.apple.com/125721264">rdar://125721264</a>

Reviewed by Said Abou-Hallawa.

Protecting the ImageFrameWorkQueue is not sufficient because the RequestQueue can be
deallocated separately.  Restore the protection of the queue.

* LayoutTests/fast/images/protected-image-queue-expected.txt: Added.
* LayoutTests/fast/images/protected-image-queue.html: Added.
* Source/WebCore/platform/graphics/ImageFrameWorkQueue.cpp:
(WebCore::ImageFrameWorkQueue::start):

Canonical link: <a href="https://commits.webkit.org/276906@main">https://commits.webkit.org/276906@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ab2c4213c4380c06fec335eb47c17387ece89596

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46094 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25224 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48679 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48765 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42134 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48401 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29546 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22666 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/37689 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46672 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22315 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39769 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18881 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19714 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40867 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4138 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42413 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41217 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50592 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21090 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17607 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/44879 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22390 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/39927 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/43782 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/10215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22749 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6423 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22084 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/ImageFrameWorkQueue.cpp:
```diff
@@ -65,8 +65,9 @@ void ImageFrameWorkQueue::start()
     m_workQueue = WorkQueue::create("org.webkit.ImageDecoder", WorkQueue::QOS::Default);
 
     m_workQueue->dispatch([protectedThis = Ref { *this }, protectedWorkQueue = Ref { *m_workQueue }, protectedSource = this->protectedSource(), protectedDecoder = Ref { *decoder }] () mutable {
+        Ref protectedFrameRequestQueue { protectedThis->requestQueue() };
```
Should not be this in the lambda capture above as the original code before the refactoring:

```
..., protectedRequestQueue = Ref { requestQueue() }, ...
```

achristensen07 committed:
REGRESSION(276827@main) Retain frame request queue in ImageFrameWorkQueue::start
https://bugs.webkit.org/show_bug.cgi?id=271979
rdar://125721264

Reviewed by Said Abou-Hallawa.

Protecting the ImageFrameWorkQueue is not sufficient because the RequestQueue can be
deallocated separately.  Restore the protection of the queue.

* LayoutTests/fast/images/protected-image-queue-expected.txt: Added.
* LayoutTests/fast/images/protected-image-queue.html: Added.
* Source/WebCore/platform/graphics/ImageFrameWorkQueue.cpp:
(WebCore::ImageFrameWorkQueue::start):

Canonical link: https://commits.webkit.org/276906@main
